### ISSUE TITLE
Role of batch_size and mini_batch_size in CrossEncoder

### ISSUE THREAD

**Issue by andreawynn** [UKPLab/sentence-transformers#3400]
Hello! I am trying to fine-tune the Qwen3-Reranker-4B-seq-cls model using the CrossEncoderTrainer class: 

```model = CrossEncoder(model_name_or_path=model_path_w_commit_id)

    loss = CachedMultipleNegativesRankingLoss(
        model=model,
        num_negatives=num_negatives,
        mini_batch_size=train_mini_batch_size,
    )

args = CrossEncoderTrainingArguments(
        ...
        per_device_train_batch_size=train_batch_size,
        per_device_eval_batch_size=train_batch_size,
        ...
    )

trainer = CrossEncoderTrainer(
        model=model,
        args=args,
        train_dataset=train_dataset['train'],
        eval_dataset=train_dataset['test'],
        loss=loss,
    )```

I am able to run this with a train_batch_size=64, train_mini_batch_size=16, and num_negatives=5 with the Qwen3-Reranker-0.6B-seq-cls model; however, I am not able to run the same code with the 4B model even when I set train_batch_size=8, train_mini_batch_size=4, and num_negatives=1. 

I would like to clarify the role of the train_batch_size and train_mini_batch_size parameters to make sure I'm not missing something - it seems to me that the above code should also work fine with the 4B model, especially with the significant reduction in batch size and num_negatives in the loss. Any help would be appreciated!

**Comment by dinhvietcuong** [comment]
I am interested in fine tuning `Qwen3-Reranker-0.6B-seq-cls` model using CrossEncoderTrainer as well.

I just wander if `CrossEncoderTrainer` is currently suitable to fine tune this `Qwen3-reranker` because as far as I understand, `Qwen3-reranker` is fine tuned / optimised with LLM (decoder side) rather than with embedding side. It would be nice if someone could enlighten me on this.

**Comment by tomaarsen** [comment]
Hello!

Indeed, the original Qwen3-Reranker models are trained as decoders that generate one of 2 tokens. The likelihood for generating each is used as the score. I've been able to fully convert this to a Sequence Classification problem with no change in performance/behaviour. This means that the models from [this collection](https://huggingface.co/collections/tomaarsen/qwen3-rerankers-converted-to-sequence-classification-684ff7b7db82ecfed889d310) should be finetunable, but I haven't tried it myself.

On to the main question:
Let's first define what each of the parameters that you mention do, so we're all on the same page:
* `num_negatives` in `CachedMultipleNegativesRankingLoss`: The (Cached) MultipleNegativesRankingLoss is also known as the InfoNCE loss in the literature, and it will essentially increase the score for query-correct_answer pairs while decreasing the score for query-incorrect_answer pairs. You can provide incorrect answers by supplying the trainer with more than 2 dataset columns (also called adding hard negatives), but the loss will also (if you wish) randomly choose "other answers" from the same batch to use as incorrect answers. The loss assumes that arbitrary answers are always incorrect for your question, but they're often so unrelated that they're a bit easy (also called random negatives or easy negatives).

    In short: higher `num_negatives` increases the number of query-answer pairs that the model has to perform inference for, as it has to compute scores for each pair during training.
* `per_device_train_batch_size` in `CrossEncoderTrainingArguments`: This is the "global" batch size, i.e. how many samples from your training/evaluation dataset is passed to the loss in one go. If you like optimizing your learning rate based on your batch size, this is the batch size that you'll use. This influences the loss value and the learning, and normally also the memory, *but* not with the `CachedMultipleNegativesRankingLoss`. This loss is special, as it can effectively mini-batch up the `per_device_train_batch_size` samples into several mini-batches, and the memory usage will be based on the size of the mini-batch instead. Speaking of which...
* `train_mini_batch_size` in `CachedMultipleNegativesRankingLoss`: This subdivides the `per_device_train_batch_size` samples into smaller mini-batches. Changing this value does *not* impact the loss value or how well your model trains, but it does define the memory usage. Lower `train_mini_batch_size` => lower memory usage. Also, because of this mini-batching, the `num_negatives` value does not influence the memory usage. You could put this higher, although it'll definitely be slower.

So, your values of train_batch_size=8, train_mini_batch_size=4, and num_negatives=1 aren't very strange, although you can easily keep the per_device_train_batch_size at 64.

When moving from the 0.6B model & training arguments to the 4B model with the different training arguments, you do have 4x smaller mini-batches (lowering memory usage), but you have to also consider that the 4B model is 6.5x larger, *and* you still have to keep the full 4B model in memory alongside the optimizer weights. The memory requirement of keeping the training samples in memory is only a piece of the puzzle.

My recommendation would be to experiment with a `train_mini_batch_size=1` and see if you can train a model with that. Also make sure that you're using `bf16=True` or `fp16=True` in the training arguments. If that still OOMs, then I think there's not much that can be done.

If that doesn't OOM, then that would be a way to finetune still. However, personally I've actually gotten the best results with the `BinaryCrossEntropyLoss` ([blogpost](https://huggingface.co/blog/train-reranker#trainer)) instead of the `CachedMultipleNegativesRankingLoss`. That one doesn't have any loss function that does mini-batching, so you'll just have to lower the `per_device_train_batch_size` a lot, but the `BinaryCrossEntropyLoss` doesn't get stronger from bigger batches, so it shouldn't matter.

Hope this helps!

- Tom Aarsen

**Comment by dinhvietcuong** [comment]
Thanks for your explanation, Tom. Will try this out in the coming time and let you know if it is working
